### PR TITLE
Future-proof Redis key changes

### PR DESCRIPTION
**Pull request by klaidliadon**
The QuotaControl system works using Redis for communication. Clients try to get keys and counter from it, and when they miss they ask the QuotaControl server to populate those values.

Right now builder prepares Redis usage counter for the client using this function:

https://github.com/0xsequence/quotacontrol/blob/09bc568f098d09f2e8256be290493721060cd6b7/client.go#L165

The client uses the same function to generate the key, and try to fetch the value from Redis using that key.

If for any reason builder and clients go out of sync, and have different versions of the function, the parts are not able to communicate through Redis.
The QuotaControl server puts the value in one key while the client is expecting it in another key, and the client times out. This happened once in the past.

I propose a small change with this PR: when asking the server to populate Redis we can provide which key to set:

https://github.com/0xsequence/quotacontrol/blob/09bc568f098d09f2e8256be290493721060cd6b7/client.go#L197

When there's no key, we generate with the usual format, otherwise we trust the client.

This will prevent the version sync issue all together.

Considering that this methods are used only in service to service communication, there should be no security concern. The fact that the key is optional, makes this retro compatible.

### CODE REVIEW

**Comment by pkieltyka** [comment]
lets skip it for this year and discuss next year

**Comment by klaidliadon** [comment]
Before:
- Client (e.g. indexer) tells the builder to populate the usage counter for a certain Access Key. 
- Builder puts the data in a key obtained using the `keyGen` function (server version)
- Client checks the redis using key obtained with `keyGen` (client version)

If client and server version of `keyGen` are different client will never manage to get the redis counter because it's looking at the wrong key.

After:
- client (e.g. indexer) tells the builder to populate the usage counter for x and passes the redis key obtained with `keyGen` (client version).
- Builder puts the data in a redis key passed by client, if no redis key is specified it uses `keyGen` (server version) top generfate it.
- client checks redis using key obtained with `keyGen` (client version)

TLDR: the client passes the redis key where it expect the counter to be to the server, so that if we update the Server keyFunc, older versions of the client will keep working.

**Comment by pkieltyka** [comment]
@klaidliadon would there be a security concern if a client tries to override the wrong key, etc..?

**Comment by klaidliadon** [comment]
The endpoint is only used in service to service communication, the access is restricted to service JWTs, so security should not be a concern.

**Comment by klaidliadon** [comment]
TLDR:

#### Before

- **client**: "I need redis usage"
- **server**: *puts the usage on redis key `x`*
- **client**: *gets the usage from redis key `x`*

If client and server don't agree on what `x` should be, we're in trouble.

#### After

- **client**: "I need redis usage in the redis key `x`"
- **server**: *puts the usage on redis key `x`*
- **client**: *gets the usage from redis key `x`*

Problem solved. No more sync issues

**Comment by VojtechVitek** [comment]
Closing per https://horizon-games-team.slack.com/archives/C06J600N7FT/p1717526673377039?thread_ts=1717502466.506099&cid=C06J600N7FT

Screen Shot 2024-07-03 at 12 17 57 PM

Feel free to reopen.
